recirq/otoc/utils.py

```diff
@@ -0,0 +1,253 @@
+                    data: np.ndarray,
+                    *,
+                    num_qubits: int = 2,
+                    add_offset: bool = False
+                    ) -> Tuple[float, np.ndarray, np.ndarray]:
+    """Obtaining a Pauli error rate from an empirical decay curve.
+
+    Args:
+        num_cycle_range: The number of cycles in an error-benchmarking
+            experiment (e.g. XEB).
+        data: Fidelity at each given cycle. Must have the same dimension as
+            num_cycle_range.
+        num_qubits: Number of qubits involved in the experiment.
+        add_offset: Whether the fitted exponential decay is forced to asymptote
+            to 0 or a non-zero offset.
+
+    Returns:
+        pauli_error: The average Pauli error per cycle from the fit.
+        x_vals: A list of 200 evenly spaced values between 0 and
+            max(num_cycle_range).
+        y_vals: The fitted fidelity for each value in x_vals.
+    """
+    f_s = 1.0e2
+
+    def _exp_decay_with_offset(length: np.ndarray, err: float,
+                               a_coeff: float, b_coeff: float) -> np.ndarray:
+        p = 1.0 - err / (1.0 - 4 ** (-num_qubits)) / f_s
+        return a_coeff * np.power(p, length) + b_coeff
+
+    def _exp_decay_no_offset(length: np.ndarray, err: float,
+                             a_coeff: float) -> np.ndarray:
+        return _exp_decay_with_offset(length, err, a_coeff, 0.0)
+
+    fit_fun = _exp_decay_with_offset if add_offset else _exp_decay_no_offset
+    params_fitted, _ = curve_fit(fit_fun, num_cycle_range, data, ftol=1e-2)
+    pauli_error = float(params_fitted[0] / f_s)
+    x_vals = np.linspace(0.0, max(num_cycle_range), 200)
+    y_vals = fit_fun(x_vals, *params_fitted)
+    return pauli_error, x_vals, y_vals
+
+
+def bits_to_probabilities(all_qubits: Sequence[Tuple[int, int]],
+                          subsys_qubits: Sequence[Tuple[int, int]],
+                          bits: np.ndarray
+                          ) -> np.ndarray:
+    """Converts a collection of bit-strings into a probability distribution.
+
+    Args:
+        all_qubits: A list of Tuples representing the (row, column) locations
+            of all qubits that are measured. The qubits are assumed to be on
+            a 2D grid.
+        subsys_qubits: The qubits whose joint probabilities are to be
+            computed. They should be a subset of all_qubits.
+        bits: a 2D np.array where each row represents a bit-string (with the
+            same length as all_qubits) and each column represents a trial.
+
+    Returns:
+        The probabilities of all possible bit-strings for subsys_qubits,
+        ordered according to the big-endian notation.
+    """
+    num_qubits = len(subsys_qubits)
+    indices = [all_qubits.index(q) for q in subsys_qubits]
+    bits_red = np.asarray(bits[:, indices], dtype=int)
+    for i in range(num_qubits):
+        bits_red[:, i] *= 2 ** (num_qubits - i - 1)
```

Comment:
what are you doing here? cant you just do `np.bincount(bits_red)` and be done with it?

Reply:
So this is computing the probability distribution from the bit-string in `bits_red`. It first turns each bit-string in `bits_red` into a base-10 number, then counts the number of occurrences for any base-10 number from 0 to 2**num_qubits.

Feeding bits_red directly to np.bincount would cause an error since bits_red is a 2D array (each row is a bit-string).

Reply:
Yeah. I think instead of

```python
bits = list(np.sum(bits_red, axis=1))
p_reduced = [bits.count(i) for i in range(2 ** num_qubits)]
```

You could use bincount in the second line (and not turn it to a list in the first line). This will likely be faster, but may not make a difference if this function isn't in a hot path